### PR TITLE
Standardize CLI tests with bin tester package 

### PR DESCRIPTION
**Pull request by scalvert**
This PR is BIG, but contains _no functional changes_ to the CLI itself. The changes are isolated to the test infra, and achieves a number of things:

1. Standardizes how we invoke tests for the bin, removing all other variants of bin test helpers
2. Converts all fixturify related APIs to async, ensuring we're correctly utilizing the new version of `fixturify-project`

### CODE REVIEW

**Comment by NullVoxPopuli** [comment]
Watching this, as this is a blocker for monorepo support, and it appears the entire v4 version of template-lint won't be compatible with monorepos (because my pr needs tests, and the tests need this pr)

Thanks for updating all the things! 🥳

**Comment by scalvert** [comment]
I'm not 100% sure I can, since bin-tester depends on some packages that are node 14 up, specifically vitest (<- you know what I'm talking about @NullVoxPopuli!). 

Let me see if I can figure things out and will update this accordingly.

**Comment by rwjblue** [comment]
It's probably find to just wait for Node 12 EOL in a couple weeks. 

> unblock @NullVoxPopuli critical monorepo plugin fix in ember-template-lint v4

How is this PR (which is just an internal testing refactor) blocking @NullVoxPopuli's monorepo fixes? They seem unrelated, if that PR is based on this one (for it's ease of testing) it can be updated to do what we do in the tests *before* this PR. There is no need for them to be coupled as far as I can tell.

**Comment by rwjblue** [comment]
> Watching this, as this is a blocker for monorepo support

@NullVoxPopuli - This statement makes no sense to me, can you explain it?

**Comment by NullVoxPopuli** [comment]
> @NullVoxPopuli - This statement makes no sense to me, can you explain it?

I was told that my PR needs to use fixturify-project to create a monorepo in the tests, which needed APIs in  fixturify-project@v4+, which was only partially migrated to, iirc -- and what this PR completes.

 (there are existing tests that I can migrate once the project is ready for that)
 
 _however_
 
 @bmish declared that
 
 > Tests for this should hopefully be unblocked now that this has been merged:
 
 _however however_, in discord,
 
 me: 
 > I need to rebase that PR, because I think someone got fixturify project upgraded? (so I can add more reasonable tests now?)
 
 @scalvert 
 > well fixturify-project was only partially updated
 > that's partly what my PR above fixes
 > but also, i just released a new fixturify-project version due to a bug in 4.1.1
 > and have subsequently upgraded it in my bin tester PR
 > so i think we should try to get that merged soon
 
 --------------
 
 So I interpreted that as this PR, #2437, being a blocker to mine -- if that's not the case, I'm happy to update things in my PR (#2359)

**Comment by bmish** [comment]
By unblocking these changes:

1. Windows support: https://github.com/ember-template-lint/ember-template-lint/pull/2339
2. Monorepo support with plugins: https://github.com/ember-template-lint/ember-template-lint/pull/2359

I was just referring to ensuring we can properly test them, and it seemed that the ability to conveniently write tests for them was blocked by some stalled test dependency upgrades.

Since these issues can require complex reproductions, I just want to ensure our tests actually cover these scenarios, so we can be confident that we fully fix them and that we don't end up breaking them again.

**Comment by rwjblue** [comment]
> I was just referring to ensuring we can properly test them, and it seemed that the ability to conveniently write tests for them was blocked by some stalled test dependency upgrades.

Gotcha. See https://github.com/ember-template-lint/ember-template-lint/pull/2471 and https://github.com/ember-template-lint/ember-template-lint/pull/2473.

**Comment by scalvert** [comment]
@bmish @rwjblue I added backwards compat support for Node 12 to `@scalvert/bin-tester`. I've also rebased this branch on top of master, so it includes @rwjblue's new monorepo tests using bin-tester too (I had to [add support for dynamic `binPath` options passed to bin-tester](https://github.com/scalvert/bin-tester/pull/40)).

I think once CI passes we should get this merged.

**Comment by scalvert** [comment]
I'll revert the package.json change to `engines` that removes Node 12 support.

**Review by bmish — APPROVED** [review]
Code changes look good. Thanks for removing the breaking change.

**Comment by scalvert** [comment]
Once I get CI back up and running I'll merge. Thanks for reviewing!

File: .github/workflows/ci.yml
Comment (bmish) [inline .github/workflows/ci.yml:57]: Why are we adding this exception for to ignore the engine check for Node 12? That defeats the check to ensure we remain compatible with Node 12.

It looks like @scalvert/bin-tester has a dependency that is not compatible with Node 12. That is where I think we should fix the issue.

Comment (scalvert) [reply .github/workflows/ci.yml:57]: Because firstly it's a devDependency, not a dependency, so it's not material whether Node 12 is supported in that dependency or not. The incompatibility is in `fixturify-project`, and that's not going to be downgraded since it was explicitly updated to make new APIs available to `bin-tester`, which in turn are required by `ember-template-lint`. 

Adding this escape hatch to CI simply avoids this issue for our tests, but doesn't negatively impact any code destined for the published package. This coupled with the short duration that we have to support Node 12 felt like an acceptable tradeoff.

Comment (scalvert) [reply .github/workflows/ci.yml]: Ya I appreciate what you're saying, @bmish. I also appreciate all the work you've done on the repo in the last while!!

I do agree there's some potential risk associated with this. I feel like it's inline with the risk associated with skipping those two test files for Node 16 - it's not ideal, but intended to be short-lived. 

I'll see if I can get CI back up and running, and for now I've separated Node 12 into its own job so that it can be easily removed once we remove it.

Comment (bmish) [reply .github/workflows/ci.yml]: Sounds good, thanks!

**Pull request merged**
